feat(goal-invitations): добавлен функционал управления приглашениями …

SergeySenin opened the pull request:
feat(goal-invitations): добавлен функционал управления приглашениями к целям.

  - Созданы соответствующие DTO (GoalInvitationDto, InvitationFilterDto) к сущностям для работы с запросами.

- Реализованы фильтры для обработки и фильтрации приглашений:
  - Создан пакет filter.goal с фильтрами:
    - GoalFilter: интерфейс для проверки условий на цель.
    - Реализованы фильтры GoalExistsFilter, UserHasActiveGoalsFilter, UserNotWorkingOnGoalFilter.
  - Создан пакет filter.goal.invitation с фильтрами:
    - InvitationFilter: интерфейс для фильтрации приглашений.
    - Реализованы фильтры InviterNamePatternFilter, InvitedNamePatternFilter, 
    InviterIdFilter, InvitedIdFilter, RequestStatusFilter.

- Добавлен GoalInvitationService:
  - Методы для создания, принятия, отклонения и получения списка приглашений:
    - createInvitation: создание приглашения с проверкой условий.
    - acceptGoalInvitation: принятие приглашения с проверкой условий.
    - rejectGoalInvitation: отклонение приглашения с проверкой условий.
    - getInvitations: получение списка приглашений с поддержкой фильтрации.

- Добавлен GoalInvitationMapper:
  - Маппинг между GoalInvitation и GoalInvitationDto.

- Добавлен GoalInvitationController:
  - Эндпоинты для создания, принятия, отклонения и получения списка приглашений.

  Добавлен класс-исключение DataValidationException.

SergeySenin committed:
feat(goal-invitations): добавлен функционал управления приглашениями к целям.

  - Созданы соответствующие DTO (GoalInvitationDto, InvitationFilterDto) к сущностям для работы с запросами.

- Реализованы фильтры для обработки и фильтрации приглашений:
  - Создан пакет filter.goal с фильтрами:
    - GoalFilter: интерфейс для проверки условий на цель.
    - Реализованы фильтры GoalExistsFilter, UserHasActiveGoalsFilter, UserNotWorkingOnGoalFilter.
  - Создан пакет filter.goal.invitation с фильтрами:
    - InvitationFilter: интерфейс для фильтрации приглашений.
    - Реализованы фильтры InviterNamePatternFilter, InvitedNamePatternFilter, InviterIdFilter, InvitedIdFilter, RequestStatusFilter.

- Добавлен GoalInvitationService:
  - Методы для создания, принятия, отклонения и получения списка приглашений:
    - createInvitation: создание приглашения с проверкой условий.
    - acceptGoalInvitation: принятие приглашения с проверкой условий.
    - rejectGoalInvitation: отклонение приглашения с проверкой условий.
    - getInvitations: получение списка приглашений с поддержкой фильтрации.

- Добавлен GoalInvitationMapper:
  - Маппинг между GoalInvitation и GoalInvitationDto.

- Добавлен GoalInvitationController:
  - Эндпоинты для создания, принятия, отклонения и получения списка приглашений.

- Реализованы проверки и валидация:
  - Проверка на существование цели.
  - Проверка на количество активных целей пользователя.
  - Проверка на наличие пользователя в участниках цели.

  Добавлен класс-исключение DataValidationException.

Rhanan1233 commented on src/main/java/school/faang/user_service/controller/goal/GoalInvitationController.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final GoalInvitationService goalInvitationService;
+
+    @Autowired
```
Можно добавить аннотацию lombok @RequiredArgsConstructor и убрать конструктор

SergeySenin committed:
@RequiredArgsConstructor.

SergeySenin committed:
Реализация и оптимизация тестов для GoalInvitationService.

- Добавлены и структурированы тесты для методов GoalInvitationService:
  - testCreateInvitationFailsForSameUser: проверяет корректность обработки исключений при попытке создать приглашение с одинаковым inviterId и invitedUserId.
  - testCreateInvitationSuccess: проверяет успешное создание приглашения.
  - testAcceptGoalInvitationSuccess: проверяет успешное принятие приглашения.
  - testRejectGoalInvitationSuccess: проверяет успешный отказ от приглашения.
  - testInvalidInvitationException: проверяет обработку исключения для несуществующего приглашения.

- Оптимизированы вспомогательные методы:
  - createTestInvitationDto: для создания DTO с заданными параметрами.
  - createTestInvitation: для создания объекта GoalInvitation с тестовыми данными.

- Настроены моки с использованием Mockito для обеспечения изоляции тестов и упрощения поддержки.

- Исправлены потенциальные ошибки с null значениями, в частности, устранена ошибка при маппинге GoalInvitationDto -> GoalInvitation.

SergeySenin committed:
Дополнен GoalInvitationController.
Добавлен GoalInvitationControllerTest.
Сделаны наброски для пакета filter.goal.

SergeySenin committed:
Рефакторинг тестов.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/goal/GoalInvitationController.java:
```diff
@@ -0,0 +1,41 @@
+    @PostMapping("/create")
+    public ResponseEntity<String> createInvitation(@RequestBody GoalInvitationDto goalInvitationDto) {
+        goalInvitationService.createInvitation(goalInvitationDto);
+        return ResponseEntity.ok("Invitation created successfully");
```
Здесь лучше прям dto возвращать. Если бы была ui, то там могли отображаться статусы. При текущем подходе мы даже не узнаем какой ИД был присвоен, статус и возможно другая полезная для пользователя информация

SnurnicynKonstantin replied:
И второй момент. /create не обязательно писать. Обычно подразумевается, что POST для создания, GET для получения...

SnurnicynKonstantin replied:
То же относится к методам ниже

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/controller/goal/GoalInvitationController.java:
```diff
@@ -0,0 +1,41 @@
+
+    private final GoalInvitationService goalInvitationService;
+
+    @PostMapping("/create")
```
И второй момент. /create не обязательно писать. Обычно подразумевается, что POST для создания, GET для получения...

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/goal/GoalInvitationService.java:
```diff
@@ -0,0 +1,95 @@
+            throw new InvalidInvitationException("Inviter and invited user cannot be the same.");
+        }
+
+        validateFilters(invitation.getGoal(), inviterId, invitedUserId);
```
Тут лучше переименовать например `goalValidation`. По сути здесь нет фильтрации, только валидация целей

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/filter/goal/InvitedNameFilterTest.java:
```diff
@@ -0,0 +1,76 @@
+package school.faang.user_service.filter.goal;
+/*
```
Эти тесты намеренно закоментировал? Если не нужны, может удалить?

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/goal/GoalInvitationServiceTest.java:
```diff
@@ -0,0 +1,135 @@
+        GoalInvitationDto testInvitationDto = createTestInvitationDto(200L);
+        GoalInvitation testInvitation = createTestInvitation();
+
+        when(goalInvitationMapper.toEntity(testInvitationDto)).thenReturn(testInvitation);
```
Маппер не нужно мокать. От должен по честному через себя пропускать объекты и трансформировать.

В юнит-тестах моки используются для замены реальных зависимостей тестируемого класса или метода на их "поддельные" реализации. Это позволяет изолировать тестируемый код и исключить влияние внешних факторов, таких как базы данных, сторонние API, файловая система или сложные вычисления.

Мокирование мапперов приводит к замене их поведения заглушками (when(...).thenReturn(...)), что скрывает ошибки в самой логике маппинга. Например, если в маппере произошла ошибка (добавилось поле или удалилось), тесты не обнаружат её, так как мок полностью игнорирует оригинальную реализацию.

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/goal/GoalInvitationServiceTest.java:
```diff
@@ -0,0 +1,135 @@
+    private GoalInvitationService goalInvitationService;
+
+    @BeforeEach
+    void setUp() {
```
Можно так создать сервис с работающим маппером
<img width="981" alt="image" src="https://github.com/user-attachments/assets/7536100b-5d4e-4454-820f-72386f41b09c" />

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/goal/GoalInvitationServiceTest.java:
```diff
@@ -0,0 +1,135 @@
+
+    @BeforeEach
+    void setUp() {
+        MockitoAnnotations.openMocks(this);
```
Можешь эту строчку удалить и вместо нее использовать аннотацию `@ExtendWith(MockitoExtension.class)`

SnurnicynKonstantin reviewed (changes requested):
Все аккуратно логично, супер.
Мелкие комменты осталось доделать.

SergeySenin committed:
Рефакторинг контроллера, сервиса и тестов.
